jupyterbook_latex/events.py

```diff
@@ -0,0 +1,102 @@
+from .transforms import (
+    LatexRootDocPostTransforms,
+    LatexRootDocTransforms,
+    MystNbPostTransform,
+)
+
+if sys.version_info < (3, 9):
+    import importlib_resources as resources
+else:
+    import importlib.resources as resources
+
+logger = logging.getLogger(__name__)
+
+
+def override_latex_config(app: Sphinx, config: Config) -> None:
+    """This ``config-inited`` event overrides aspects of the sphinx latex config.
+
+    - ``latex_engine`` -> ``xelatex``
+    - ``latex_theme`` -> ``jupyterBook``
+    - appends necessary LaTeX commands to the preamble
+
+    """
+    # only allow latex builder to access rest of the features
+    config["latex_engine"] = "xelatex"
+    config["latex_theme"] = "jupyterBook"
+
+    latex_elements = cast(dict, config["latex_elements"])
+
+    # preamble to overwrite things from sphinx latex writer
+    config_preamble = (
+        latex_elements["preamble"] if "preamble" in config["latex_elements"] else ""
+    )
+
+    latex_elements["preamble"] = (
+        config_preamble
+        + r"""
+         \usepackage[Latin,Greek]{ucharclasses}
+        \usepackage{unicode-math}
+        % fixing title of the toc
+        \addto\captionsenglish{\renewcommand{\contentsname}{Contents}}
+        """
+    )
+
+
+def setup_latex_transforms(app: Sphinx) -> None:
+    """This ``builder-inited`` event sets up aspects of the extension,
+    reserved only for when a LaTeX builder is specified.
+    """
+
+    if not isinstance(app.builder, LaTeXBuilder):
+        return
+
+    # note: bold is a dynamically created function
+    from sphinx.util.console import bold  # type: ignore[attr-defined]
+
+    logger.info(
+        bold("jupyterbook-latex v%s:") + "(latex_engine='%s')",
+        __version__,
+        app.config["latex_engine"],
+    )
+
+    # Copy the class theme to the output directory.
+    # note: importlib.resources is the formal method to access files within packages
+    with resources.as_file(resources.files(theme).joinpath("jupyterBook.cls")) as path:
+        copy_asset_file(str(path), app.outdir)
+
+    # only load when myst-nb is present
+    if MystNbPostTransform.check_dependency():
+        app.add_post_transform(MystNbPostTransform)
+
+    toc_path = Path(app.confdir or app.srcdir).joinpath("_toc.yml")
+    if not os.path.exists(toc_path):
+        logger.info(
+            "Some features of this extension will work only with a jupyter-book application"  # noqa: E501
+        )
+        return
+
+    # TODO why is this necessary, I don't think this should be enforced
+    if (
+        "myst_enable_extensions" in app.config
+        and "amsmath" not in app.config["myst_enable_extensions"]
+    ):
+        logger.info("[jb-latex]: Adding 'amsmath' to myst-parser extensions")
+        app.config["myst_enable_extensions"].append(  # type: ignore[attr-defined]
+            "amsmath"
+        )
+
+    app.setup_extension("sphinx.ext.imgconverter")
+    app.add_transform(LatexRootDocTransforms)
```

Comment:
This is faulty logic, which will fail for rebuilds, if you run e.g. a HTML build then a LaTeX build; the HTML build would store the doctree (without having run `LatexRootDocTransforms`) then the latex build will take that doctree and go straight to post-transforms,
i.e. you can't have transforms that are builder specific, only post-transforms.

Reply:
can the functionality in this transform be moved to a post-transform

Reply:
@mmcky @AakashGfude was there a specific reason for splitting the "root doc restructure" into a transform and post-transform? e.g. is there certain things you know that need to happen before/after other things?

Reply:
@chrisjsewell I remember there was a reason to do this in transform because of the state of the doctree required. Will go through the code to pinpoint the exact reasons.

Reply:
I've moved all the restructuring to the post-transform, which works